src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java

```diff
@@ -618,51 +674,88 @@ public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext
+                // then take the logical-or across the expressions in infoJexls.
+                //
+                // For example, given the command line
+                // "-select AF > 0.01 -select ReadPosRankSum < -20.0"
+                // If invert-select is false, the expression is understood to be "AF > 0.01 || ReadPosRankSum < -20.0"
+                // If invert-select is true, it is "!(AF > 0.01) || !(ReadPosRankSum < -20.0)", or
+                // Note that this is not equivalent to "!(AF > 0.01 || ReadPosRankSum < -20.0)"
+
+                // Notice here that calling the match method without the genotype g leads to genotype g being set to null,
+                // which is fine since infoJexls should not refer to genotype fields (unless using vc.getGenotype(), in which case
+                // JEXL can access genotype     fields via vc)
+                if (invertLogic(VariantContextUtils.match(vc, jexl), invertSelect)){
+                    return true;
                 }
-            } catch (IllegalArgumentException e) {
-                // The IAE thrown by htsjdk already includes an informative error message ("Invalid JEXL
-                //  expression detected...")
-                throw new UserException(e.getMessage() +
-                        "\nSee https://gatk.broadinstitute.org/hc/en-us/articles/360035891011-JEXL-filtering-expressions for documentation on using JEXL in GATK", e);
             }
 
-            if (!failedJexlMatch &&
-                    (!selectRandomFraction || Utils.getRandomGenerator().nextDouble() < fractionRandom)) {
-                //remove annotations being dropped and write variantcontext
-                final VariantContext variantContextToWrite = buildVariantContextWithDroppedAnnotationsRemoved(filteredGenotypeToNocall);
-                pendingVariants.add(variantContextToWrite);
+            // ##### Apply Genotype JEXL filters #####
+            for (VariantContextUtils.JexlVCMatchExp jexl : genotypeJexls) {
+                // We separate the genotype vs INFO cases into two separate arguments because we don't want to be
+                // looping over all genotypes by default
+                for (Genotype g : vc.getGenotypes()){
+                    if (invertLogic(VariantContextUtils.match(vc, g, jexl), invertSelect)){
+                        return true;
+                    }
+                }
             }
+        } catch (IllegalArgumentException e) {
+            // The IAE thrown by htsjdk already includes an informative error message ("Invalid JEXL
+            //  expression detected...")
+            throw new UserException(e.getMessage() +
+                    "\nSee https://gatk.broadinstitute.org/hc/en-us/articles/360035891011-JEXL-filtering-expressions for documentation on using JEXL in GATK", e);
+        }
+
+        if (vc.isFullyDecoded()){
+            int d = 3; // tsato: is it ever fully-decoded?
```

Comment:
I think you can remove this debugging code :)

Reply:
Done